README.md
# DaKanji Single Kanji Recognition

This CNN can recognize a single character (Kanji, Hiragana, Katakana).
A list of all supported characters can be found [here](./labels.txt).

## Apps which use this model

| name | android | iOS | Linux | MacOS | Windows | Web |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| DaKanji | ✅ |   | ✅ | ✅ | ✅ |   |
| Kanji Graph | ✅ |   |   |   |   |   |

## Training

To generate the data necessary to train this CNN, the [single_kanji_data_gen notebook](single_kanji_data_gen.ipynb) is used.
The training can than be done with the [single_kanji_cnn_training notebook](single_kanji_cnn_training.ipynb).

## Inference

In the releases section pretrained model weights can be found. Also a TensorFlow lite model is available.

**Input:**
The input should be a *grayscale* image of *any size*.

**Output:**
A one-hot-vector containing the class probabilities (lines up with `labels.txt`).

## Setup development environment

install all dependencies:

``` python
python -m pip install wheel
python -m pip install -r requirements.txt
```

Now you should follow model specific setup steps.
For this look at the README for the model you are interested in.

## Usage in your software

I put lots of effort and time into developing this model and hope that it can be used in many apps.
If you decide to use this machine learning model please give me credit like:
`Character recognition powered by machine learning from CaptainDario (DaAppLab)`
It would also be nice if you open an issue and tell me that you are using this model.
Than I would add your software to the [test](#apps-which-use-this-model)

## Credits

The data on which the neural network was trained on was kindly provided by [ETL Character Database](http://etlcdb.db.aist.go.jp/obtaining-etl-character-database)

[The KanjiVG dataset](https://kanjivg.tagaini.net/) was used to generate "handwritten" kanjis
Papers:

* [EfficientNet: Rethinking Model Scaling for Convolutional Neural Networks](https://arxiv.org/abs/1905.11946)
* [Recognizing Handwritten Japanese Characters Using Deep Convolutional Neural Networks](http://cs231n.stanford.edu/reports/2016/pdfs/262_Report.pdf)
* [A neural framework for online recognition of handwritten Kanji characters](https://www.researchgate.net/publication/327893142_A_neural_framework_for_online_recognition_of_handwritten_Kanji_characters)
* [Online Handwritten Kanji Recognition Based on Inter-stroke Grammar](https://www.researchgate.net/publication/4288187_Online_Handwritten_Kanji_Recognition_Based_on_Inter-stroke_Grammar)


CHANGELOG.md
# DaKanjiRecognizer - Machine Learning : changelog


## v1.2

new Features:

- recognize:

  - all 漢字 from 漢字検定
  - ひらがな (also historical ones: ゑ, etc.)
  - カタカナ (also historical ones: ヱ, etc.)

Changes:

- handle class imbalance better
- moved fonts into root directory

-------------------------------------------------------------------------

## v 1.1

changes:

- moved from jupyter notebook to jupyter lab
- multi processing for loading the data
- data generator for feeding batches to the CNN
  - multi processed
  - image augmentation

-------------------------------------------------------------------------

## v 1.0

features:

- recognize ~3000 kanji characters
  

RandomErasing.py
import tensorflow as tf
import numpy as np



class RandomErasing(tf.keras.layers.Layer):
    """A Layer which applies Random Erasing image augmenation to its input.
    """

    def __init__(self, name="random_erasing", 
        probability : float = 0.5, sl : float = 0.02, sh : float = 0.4, r1 : float = 0.3, method : str = 1,
        train_only = True, **kwargs):
        """
        Args:
            img         : 3D Tensor data (H,W,Channels) normalized value [0,1]
            probability : The probability that the operation will be performed.
            sl          : min erasing area
            sh          : max erasing area
            r1          : min aspect ratio
            method      : Erasing type : 1 - ('black'), 2 - ('white') or 3 - ('random') 
        """
        super(RandomErasing, self).__init__(name=name, **kwargs)
        self.probability = probability
        self.sl = sl
        self.sh = sh
        self.r1 = r1
        self.method = method
        self.train_only = train_only

    def call(self, input):
        if(tf.keras.backend.learning_phase() or not self.train_only):
            input = tf.map_fn(self.random_erasing, input)
        return input

    def get_config(self):
        config = super(RandomErasing, self).get_config()
        return config

@tf.function
def random_erasing(self, img) -> tf.Tensor:
    '''
    Method that performs Random Erasing from "Random Erasing Data Augmentation" by Zhong et al..

    Args:
        img : the image on which Randomerasing should be applied.

    Returns:
        The augmented image.
    '''  

    probability = 0.3
    sl = 0.02
    sh = 0.4
    r1 = 0.3
    method = 1
    
    if tf.random.uniform([]) > probability:
        return tf.convert_to_tensor(img)

    img_width    = img.shape[1]
    img_height   = img.shape[0]
    img_channels = img.shape[2]

    area = img_height * img_width

    target_area = tf.random.uniform([],minval=sl, maxval=sh) * area
    aspect_ratio = tf.random.uniform([],minval=r1, maxval=1/r1)
    h = tf.cast(tf.math.round(tf.math.sqrt(target_area * aspect_ratio)), tf.int32)
    w = tf.cast(tf.math.round(tf.math.sqrt(target_area / aspect_ratio)), tf.int32)

    while tf.constant(True, dtype=tf.bool):
        if h > img_height or w > img_height:
            target_area = tf.random.uniform([],minval=sl, maxval=sh) * area
            aspect_ratio = tf.random.uniform([],minval=r1, maxval=1/r1)
            h = tf.cast(tf.math.round(tf.math.sqrt(target_area * aspect_ratio)), tf.int32)
            w = tf.cast(tf.math.round(tf.math.sqrt(target_area / aspect_ratio)), tf.int32)
        else:
            break

    x1 = tf.cond(img_height == h, lambda:0, lambda:tf.random.uniform([], minval=0, maxval=img_height - h, dtype=tf.int32))
    y1 = tf.cond(img_width  == w, lambda:0, lambda:tf.random.uniform([], minval=0, maxval=img_width - w, dtype=tf.int32))

    part1 = tf.slice(img, [0,0,0], [x1,img_width,img_channels]) # first row
    part2 = tf.slice(img, [x1,0,0], [h,y1,img_channels]) # second row 1

    # black
    if method == 1:
        part3 = tf.zeros((h,w,img_channels), dtype=tf.float16) # second row 2
    # white
    elif method == 2:
        part3 = tf.ones((h,w,img_channels), dtype=tf.float16)
    # random
    elif method == 3:
        part3 = tf.random.uniform((h,w,img_channels), dtype=tf.float16)

    part4 = tf.slice(img,[x1,y1+w,0], [h,img_width-y1-w,img_channels]) # second row 3
    part5 = tf.slice(img,[x1+h,0,0], [img_height-x1-h,img_width,img_channels]) # third row

    middle_row = tf.concat([part2,part3,part4], axis=1)
    img = tf.concat([part1,middle_row,part5], axis=0)

    return img   

generate_kanji_from_vg.py
import time
import os

from xml.etree import ElementTree as ET

import drawSvg as draw
from svgpathtools import parse_path, disvg
import svgpathtools
from svgpathtools.svg_to_paths import svg2paths



def kanjivg_xml_to_image(tag, strokes):
	"""
	Recursively traverses the given `tag` and appends all found strokes
	to the `strokes` list.

	Note: This function works in-place and adds all found elements to the `strokes`-list

	Args:
		tag       (str) : 
		strokes ([str]) : A nested list containing all strokes.

	"""

	# this is a radical (consists of multipe strokes) of the kanji
	if(tag.tag == 'g'):
		# iterate over all children of the radical
		for child in tag:
			# if the child of this radical is another radical
			# add a list to store all strokes belonging to this radical in
			if(child.tag == 'g'):
				path = []
				strokes.append(path)
				kanjivg_xml_to_image(child, path)
			# if it is a leave than don't add a new list but add the indivdual strokes
			else:
				kanjivg_xml_to_image(child, strokes)
	# this is a stroke
	elif(tag.tag == 'path'):
		strokes.append(tag.attrib['d'])


def strokes_to_svg(strokes):
	d = draw.Drawing(107, 107, origin='center', displayInline=False)

	segments = []
	for cnt, stroke in enumerate(strokes):
		sub_segments = parse_path(stroke)
		segments.extend(sub_segments._segments)
		
	svg = svgpathtools.Path(*segments)
	svgpathtools.disvg(svg)

	print("test")

if __name__ == "__main__":

	# read the kanjivg data base from file
	tree = ET.parse('kanjivg.xml')
	root = tree.getroot()

	# read all jis level 2 characters from file
	kanjis = None
	with open(r"E:\projects\DaKanjiRecognizerML\single_kanji_cnn\jis2_characters.txt", encoding="utf8") as f:
		kanji = f.read()
	kanjis = kanji.replace("\n", " ").split(" ")

	# iterate over all kanji xml tags
	cnt = 0
	for child in root:

		hex = child.attrib['id'].removeprefix("kvg:kanji_")
		kanji = chr(int(hex, 16))
		if(kanji == "わ"):
			# get the strokes from the database
			strokes = []
			kanjivg_xml_to_image(child[0], strokes)

			# draw the char on an image
			strokes_to_svg(strokes)

		cnt += 1

	print(cnt)


single_kanji_cnn/erasing.py
import tensorflow as tf
import numpy as np



def random_erasing(img : np.ndarray) -> tf.Tensor:
    #Motivated by https://github.com/Amitayus/Random-Erasing-TensorFlow.git
    #Motivated by https://github.com/zhunzhong07/Random-Erasing/blob/master/transforms.py
    '''
    Method that performs Random Erasing from "Random Erasing Data Augmentation" by Zhong et al..
    
    A random rectangle from the input image will be selected and mathing to `method` this part will be covered. 
    
    Args:
        img         : 3D Tensor data (H,W,Channels) normalized value [0,1]
        probability : The probability that the operation will be performed.
        sl          : min erasing area
        sh          : max erasing area
        r1          : min aspect ratio
        method      : 'black', 'white' or 'random'. Erasing type
    
    Returns:
        The augmented image. 
    '''
    
    probability : float = 0.5
    sl: float = 0.02
    sh: float = 0.4
    r1: float = 0.3
    method : str = 'black'

    
    if tf.random.uniform([]) > probability:
        return tf.convert_to_tensor(img)

    img_width    = img.shape[1]
    img_height   = img.shape[0]
    img_channels = img.shape[2]

    area = img_height * img_width

    target_area = tf.random.uniform([],minval=sl, maxval=sh) * area
    aspect_ratio = tf.random.uniform([],minval=r1, maxval=1/r1)
    h = tf.cast(tf.math.round(tf.math.sqrt(target_area * aspect_ratio)), tf.int32)
    w = tf.cast(tf.math.round(tf.math.sqrt(target_area / aspect_ratio)), tf.int32)

    while tf.constant(True, dtype=tf.bool):
        if h > img_height or w > img_height:
            target_area = tf.random.uniform([],minval=sl, maxval=sh) * area
            aspect_ratio = tf.random.uniform([],minval=r1, maxval=1/r1)
            h = tf.cast(tf.math.round(tf.math.sqrt(target_area * aspect_ratio)), tf.int32)
            w = tf.cast(tf.math.round(tf.math.sqrt(target_area / aspect_ratio)), tf.int32)
        else:
            break

    x1 = tf.cond(img_height == h, lambda:0, lambda:tf.random.uniform([], minval=0, maxval=img_height - h, dtype=tf.int32))
    y1 = tf.cond(img_width  == w, lambda:0, lambda:tf.random.uniform([], minval=0, maxval=img_width - w, dtype=tf.int32))
    
    part1 = tf.slice(img, [0,0,0], [x1,img_width,img_channels]) # first row
    part2 = tf.slice(img, [x1,0,0], [h,y1,img_channels]) # second row 1

    if method == 'black':
        part3 = tf.zeros((h,w,img_channels), dtype=tf.float16) # second row 2
    elif method == 'white':
        part3 = tf.ones((h,w,img_channels), dtype=tf.float16)
    elif method == 'random':
        part3 = tf.random.uniform((h,w,img_channels), dtype=tf.float16)
    
    part4 = tf.slice(img,[x1,y1+w,0], [h,img_width-y1-w,img_channels]) # second row 3
    part5 = tf.slice(img,[x1+h,0,0], [img_height-x1-h,img_width,img_channels]) # third row

    middle_row = tf.concat([part2,part3,part4], axis=1)
    img = tf.concat([part1,middle_row,part5], axis=0)

    return img   

single_kanji_cnn/DataGenerator.py
import numpy as np
import PIL
from PIL import Image as PImage
from PIL import ImageFilter, ImageFont, ImageDraw
import cv2
import tensorflow as tf

from fontTools.ttLib import TTFont

from typing import Tuple, List
import random



def distort_sample(img : PImage) -> Tuple[PIL.Image.Image, Tuple[int, int], Tuple[int, int]]:
    """
    Distort the given image randomly.

    Randomly applies the transformations:
        rotation, shear, scale, translate, 
    Randomly applies the filter:
        sharpen, blur, smooth

    Args:
        img: the image which should be distorted.

    Returns:
        the distorted image
        the offset where the character is placed in the image
        the size of the character
    """

    offset, scale = (0, 0), (64, 64)

    t = random.choice(["rotate", "shear", "erode", "dilate"])
    f = random.choice(["blur", "sharpen", "smooth"])

    # randomly apply transformations...
    # rotate image
    if("rotate" in t):
        img = img.rotate(random.uniform(-10, 10))
    
    # shear image
    if("shear" in t):
        y_shear = random.uniform(-0.2, 0.2)
        x_shear = random.uniform(-0.2, 0.2)
        img = img.transform(img.size, PImage.AFFINE, (1, x_shear, 0, y_shear, 1, 0))

    # sine transform
    t_img = np.array(img)
    
    A = t_img.shape[0] / random.uniform(3.0, 6.0)
    w = 2.0 / t_img.shape[1]

    shift_factor = random.choice([-1, 1]) * random.uniform(0.10, 0.2)
    shift = lambda x: shift_factor * A * np.sin(-2*np.pi*x * w)

    for i in range(t_img.shape[0]):
        t_img[:,i] = np.roll(t_img[:,i], int(shift(i)))
    
    if("erode" in t):
        t_img = cv2.erode(t_img, np.ones((2, 2), np.uint8), iterations=1)
    if("dilate" in t):
        t_img = cv2.dilate(t_img, np.ones((3, 3), np.uint8), iterations=1)
    
    img = PImage.fromarray(t_img, mode="L")

    # FILTER
    # blur
    if("blur" in f):
        img = img.filter(ImageFilter.GaussianBlur(radius=random.uniform(0.5, 1.2)))

    # sharpen
    if("sharpen" in f):
        img = img.filter(ImageFilter.SHARPEN)
        
    # smooth
    if("smooth" in f):
        img = img.filter(ImageFilter.SMOOTH)

    return img, offset, scale


def generate_images(
    amounts : List[int], kanjis : List[str], 
    fonts : List[List[str]], font_size : int = 50) -> Tuple[np.ndarray, np.ndarray]:
    """
    Generates `amount` images of the given `kanji` using given `fonts`
    while augmenting 50% of them randomly. 

    Args:
        amount    : how many samples should be created per kanji.
        kanji     : the (kanji) character which should be created
        fonts     : a list of paths to fonts which should be used 
                    to created images.

    Returns:
        Two numpy arrays.
        One containing all images and the other the matching labels.
    """

    cnt = 0
    kanji_labels = np.empty(shape=(sum(amounts)), dtype=str)
    kanji_imgs = np.zeros(shape=(sum(amounts), 64, 64, 1), dtype=np.uint8)

    
    for kanji_cnt, kanji in enumerate(kanjis):
        already_gen_cnt = sum(amounts[:kanji_cnt])
        ttfs = [ImageFont.truetype(f, font_size) for f in fonts[kanji_cnt]]

        # generate one image per font
        for cnt, font in enumerate(ttfs):

            if(cnt >= amounts[kanji_cnt]):
                break

            # make sure that the image fits in the 64x64 image
            if(font.getsize(kanji)[0] > 64):
                font = font.font_variant(size = font_size - (font.getsize(kanji)[0] - font_size))
            if(font.getsize(kanji)[1] > 64):
                font = font.font_variant(size = font_size - (font.getsize(kanji)[1] - font_size))

            # create the image
            img = PImage.new(mode="L", size=(64, 64), color=0)
            d = ImageDraw.Draw(img)
            d.text(((64 - font.getsize(kanji)[0]) // 4, (64 - font.getsize(kanji)[1]) // 4), kanji, font=font, fill=255)     

            # store the image / label in the array
            kanji_imgs[cnt + already_gen_cnt] = np.array(img).reshape(64, 64, 1)
            kanji_labels[cnt + already_gen_cnt] = kanji
        
        # append distorted versions of the images created from the fonts
        cnt += 1
        while(cnt < amounts[kanji_cnt]):
            p_img = tf.keras.preprocessing.image.array_to_img(kanji_imgs[(cnt) % len(ttfs) + already_gen_cnt])
            kanji_imgs[cnt+already_gen_cnt]   = np.array(distort_sample(p_img)[0]).reshape(64, 64, 1)
            kanji_labels[cnt+already_gen_cnt] = kanji
            cnt += 1
        
    return kanji_imgs, kanji_labels

def check_font_char_support(fonts : List[str], char : str):
    """ Check which of the given fonts supports the given char.
    """

    element = (char, [])

    for font in fonts:
        f = ImageFont.truetype(font, 64)
        for table in TTFont(font)['cmap'].tables:
            if ord(char) in table.cmap.keys():
                # create an image and draw the character
                img = PImage.new(mode="L", size=(64, 64), color=0)
                d = ImageDraw.Draw(img)
                d.text((0, 0), char, font=f, fill=255)
                img = np.array(img).reshape(64, 64, 1)    

                if(np.any(img)):
                    element[1].append(font)
                break

    return element

CNN_kanji/jupyter/training.py

#std lib
import os
import random
import math
import random
import multiprocessing as mp
import gc
import datetime

#reading the dataset
from etldr.etl_data_reader import ETLDataReader
from etldr.etl_character_groups import ETLCharacterGroups
from etldr.etl_data_names import ETLDataNames

#data handling
from PIL import Image as PImage
from PIL import ImageFilter
import numpy as np


#plotting/showing graphics
import matplotlib.pyplot as plt
from IPython.display import Image
#define a font to show japanese characters in matplotlib figures
import matplotlib.font_manager as fm
font = fm.FontProperties(fname=os.path.join("font", "NotoSerifCJKjp-Regular.otf"), size=20)



shared_dict = {}


def distort_sample(img : Image) -> (Image, [int], [int]):
    """
    Distort the given image randomly.

    Randomly applies the transformations:
        rotation, shear, scale, translate, 
    Randomly applies the filter:
        sharpen, blur, smooth

    Returns the distorted image.
    """

    offset, scale = (0, 0), (64, 64)

    t = random.choice(["sine", "rotate", "shear", "scale"])
    f = random.choice(["blur", "sharpen", "smooth"])

    # randomly apply transformations...
    # rotate image
    if("rotate" in t):
        img = img.rotate(random.uniform(-15, 15))
    
    # shear image
    if("shear" in t):
        y_shear = random.uniform(-0.2, 0.2)
        x_shear = random.uniform(-0.2, 0.2)
        img = img.transform(img.size, PImage.AFFINE, (1, x_shear, 0, y_shear, 1, 0))
    
    # scale and translate image
    if("scale" in t):
        #scale the image
        size_x = random.randrange(25, 63)
        size_y = random.randrange(25, 63)
        scale = (size_x, size_y)
        offset = (math.ceil((64 - size_x) / 2), math.ceil((64 - size_y) / 2))
        img = img.resize(scale)

        # put it again on a black background (translated)
        background = PImage.new('L', (64, 64))
        trans_x = random.randrange(0, math.floor((64 - size_x)))
        trans_y = random.randrange(0, math.floor((64 - size_y)))
        offset = (trans_x, trans_y)
        background.paste(img, offset)
        img = background
    
    if("sine" in t):
        t_img = np.array(img)

        A = t_img.shape[0] / 3.0
        w = 2.0 / t_img.shape[1]

        shift_factor = random.choice([-1, 1]) * random.uniform(0.15, 0.2)
        shift = lambda x: shift_factor * A * np.sin(-2*np.pi*x * w)

        for i in range(t_img.shape[0]):
            t_img[:,i] = np.roll(t_img[:,i], int(shift(i)))

        img = PImage.fromarray(t_img)


    # blur
    if("blur" in f):
        img = img.filter(ImageFilter.GaussianBlur(radius=random.uniform(0.5, 1.2)))

    # sharpen
    if("sharpen" in f):
        img = img.filter(ImageFilter.SHARPEN)
        
    # smooth
    if("smooth" in f):
        img = img.filter(ImageFilter.SMOOTH)

    return img, offset, scale

def get_model(name : str):
    _model = tf.keras.models.Sequential([
        tf.keras.layers.Conv2D(input_shape=(64, 64, 1), kernel_size=3, activation='relu', filters=32, name="conv2D_1_2_input"),
        tf.keras.layers.Conv2D(kernel_size=3, activation='relu', filters=32, name="conv2D_1_1"),
        tf.keras.layers.MaxPool2D(pool_size=(2, 2), strides=2, name="maxpool_1"),

        tf.keras.layers.Conv2D(kernel_size=3, activation='relu', filters=32, name="conv2D_2_1"),
        tf.keras.layers.Conv2D(kernel_size=3, activation='relu', filters=32, name="conv2D_2_2"),
        tf.keras.layers.MaxPool2D(pool_size=(2, 2), strides=2, name="maxpool_2"),

        tf.keras.layers.Conv2D(kernel_size=3, activation='relu', filters=64, name="conv2D_3_1"),
        tf.keras.layers.Conv2D(kernel_size=3, activation='relu', filters=64, name="conv2D_3_2"),
        tf.keras.layers.MaxPool2D(pool_size=(2, 2), strides=2, name="maxpool_3"),

        tf.keras.layers.Conv2D(kernel_size=3, activation='relu', filters=128, name="conv2D_4_1"),
        tf.keras.layers.MaxPool2D(pool_size=(2, 2), strides=2, name="maxpool_4"),

        tf.keras.layers.Flatten(name="flatten_1"),
        tf.keras.layers.Dropout(0.25, name="dropout_1"),
        
        tf.keras.layers.Dense(2048, name="dense_1"),
        tf.keras.layers.Dropout(0.1, name="dropout_2"),

        tf.keras.layers.Dense(2048, name="dense_2"),
        tf.keras.layers.Dropout(0.25, name="dropout_3"),

        tf.keras.layers.Dense(len(lb.classes_), name="dense_3"),

        #set the dtype to float32 for numerical stability
        tf.keras.layers.Softmax(dtype="float32", name="softmax_1_output") 
    ], name=name)
    
    return _model

def generator_initializer(_x_shared, _y_shared):
    shared_dict["x"] = _x_shared
    shared_dict["y"] = _y_shared

def generator_func(start_index, end_index, x_shape, y_shape):
    X, Y = [], []
    
    x_loc = np.frombuffer(shared_dict["x"], dtype="float16").reshape(x_shape)
    y_loc = np.frombuffer(shared_dict["y"], dtype="b").reshape(y_shape)
    
    for i in range(start_index, end_index):
        base_img = x_loc[i]
        img = PImage.fromarray(np.uint8(base_img.reshape(64, 64) * 255))
        img, *unused = distort_sample(img)

        # add transformed image
        X.append(np.array(img).reshape(64, 64, 1))
        Y.append(y_loc[i])

        # add base image
        X.append(base_img)
        Y.append(y_loc[i])
        
    return X, Y



if __name__ == "__main__":
    #ML
    import tensorflow as tf
    from tensorflow.keras.callbacks import ModelCheckpoint

    #creating one hot encodings
    from sklearn.preprocessing import LabelBinarizer
    
    class DataGenerator(tf.keras.utils.Sequence):

        def __init__(self, num_samples, batch_size, percentage, mode, processes, shuffle=True):
            self.num_samples = num_samples
            # 50% original images + 50% augmented images 
            self.batch_size = batch_size // 2
            self.percentage = percentage

            # an offset to devide the data set into test and train
            self.start_index = 0
            if(mode == "testing"):
                self.start_index = num_samples - (num_samples // 100 * percentage)
            # is this a train or a test generator
            self.mode = mode
            # how many processes should be used for this generator
            self.processes = processes
            # should the arrays be shuffled after each epoch
            self.shuffle = shuffle
            
            # a pool of processes for generating augmented data
            self.pool = mp.Pool(processes=self.processes,
                initializer=generator_initializer,
                initargs=(x_shared, y_shared))
            
        def __len__(self):
            return (self.num_samples // 100 * self.percentage) // self.batch_size

        def on_epoch_end(self):
            if(self.shuffle):
                rng_state = np.random.get_state()
                np.random.shuffle(x_np)
                np.random.set_state(rng_state)
                np.random.shuffle(y_np)
                
        def __getitem__(self, index):

            arguments = []
            slice_size = self.batch_size // self.processes
            current_batch = index * self.batch_size
            for i in range(self.processes):
                slice_start = self.start_index + (current_batch + i * slice_size)
                slice_end = self.start_index + (current_batch + (i+1) * slice_size)
                arguments.append([slice_start, slice_end, x_np.shape, y_np.shape])
            
            return_values = self.pool.starmap(generator_func, arguments)

            X, Y = [], []
            for imgs, labels in return_values:
                X.append(imgs)
                Y.append(labels)

            return np.concatenate(X).astype(np.float16), np.concatenate(Y).astype(np.float16)
    

    print("GPUs Available: ", tf.test.gpu_device_name())
    physical_devices = tf.config.list_physical_devices('GPU') 
    tf.config.experimental.set_memory_growth(physical_devices[0], True)
    tf.keras.mixed_precision.set_global_policy("mixed_float16")

    # read data
    path = "F:\data_sets\ETL_kanji"
    reader = ETLDataReader(path)
    x, y =  reader.read_dataset_whole(include=[ETLCharacterGroups.kanji], processes=16)
    print("finished loading data")

    # because the data is ordered shuffle it
    rng_state = np.random.get_state()
    np.random.shuffle(x)
    np.random.set_state(rng_state)
    np.random.shuffle(y)

    #one hot encode the labels
    lb = LabelBinarizer()
    lb.fit(y)
    o_y = lb.transform(y)

    # free the memory from the original string labels
    del(y)
    print("finished one-hot-encoding")
    print(o_y.shape)


    # create shared arrays
    memory_size_x = x.shape[0] * x.shape[1] * x.shape[2] * x.shape[3]
    x_shared = mp.RawArray("f", memory_size_x // 2)
    x_np = np.frombuffer(x_shared, dtype="float16").reshape(x.shape)
    np.copyto(x_np, x)
    del(x)

    memory_size_y = o_y.shape[0] * o_y.shape[1]
    y_shared = mp.RawArray('b', memory_size_y)
    y_np = np.frombuffer(y_shared, dtype="b").reshape(o_y.shape)
    np.copyto(y_np, o_y)
    del(o_y)
    gc.collect()
    print("finished moving data in shared memory")


    # define CNN
    #path where the model should be saved
    model_dir = os.path.join(os.getcwd(), "CNN_kanji", "model")
    #f16_model = get_model("DaKanjiRecognizer_f16")
    f16_model = tf.keras.models.load_model(os.path.join(model_dir, "tf", "checkpoints", "weights-improvement-108-0.98.hdf5"))
    #print(f16_model.output_shape)
    #f16_model.summary()
    print("Saving with base path:", model_dir)
    opt = tf.keras.optimizers.Adam(learning_rate=0.0001,
                                    beta_1=0.9,
                                    beta_2=0.999,
                                    epsilon=1e-08,)

    f16_model.compile(optimizer=opt,
                loss="categorical_crossentropy",
                metrics=['accuracy'])


    # training the CNN
    # percentage to split into test and train
    train_generator = DataGenerator(len(x_np), 1024, 80, "training", 4)
    test_generator  = DataGenerator(len(x_np), 1024, 20, "testing" , 4)
    print("finished instanciating data generator")

    #checkpoints setup
    filepath = os.path.join(model_dir, "tf", "checkpoints", "weights-improvement-{epoch:02d}-{val_accuracy:.2f}.hdf5")
    checkpoint = ModelCheckpoint(filepath, monitor='val_accuracy', verbose=1, save_best_only=True, mode='max')
    
    log_dir = os.path.join(model_dir, "logs", datetime.datetime.now().strftime("%Y%m%d-%H%M%S"))
    tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)

    callbacks_list = [checkpoint, tensorboard_callback]

    #train the model
    hist = f16_model.fit(
        x=train_generator,
        epochs=200,
        initial_epoch=109,
        validation_data=test_generator,
        max_queue_size=200,
        callbacks=callbacks_list,
        workers=1
    )

    # saving the model
    f16_model.save(os.path.join(model_dir, "tf", "trained_model"))

    # Create a float32 model with the same weights as the mixed_float16 model, so
    # that it loads into TF Lite
    tf.keras.mixed_precision.set_global_policy("float32")
    f32_model = get_model("DaKanjiRecognizer_f32")
    f32_model.set_weights(f16_model.get_weights())
    #f32_model.summary()

    #save the model as tflite
    # Convert the model
    converter = tf.lite.TFLiteConverter.from_keras_model(f32_model) # path to the SavedModel directory
    tflite_model = converter.convert()

    # Save the model.
    with open(os.path.join(model_dir, "tflite", "model.tflite"), 'wb') as f:
        f.write(tflite_model)
